{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vhtZncrG3JZ4"
      },
      "source": [
        "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/protosome/convergent_overlapping_gene_prediction/blob/main/codon_overlap.ipynb)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EtSjnKef3MTH"
      },
      "outputs": [],
      "source": [
        "!git clone https://github.com/protosome/convergent_overlapping_gene_prediction.git"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qa97Rkfz3Ur9"
      },
      "outputs": [],
      "source": [
        "%cd convergent_overlapping_gene_prediction/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Y8KuebJyBFM7"
      },
      "outputs": [],
      "source": [
        "pip install biopython"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "IDCP5PDyBRhq"
      },
      "outputs": [],
      "source": [
        "pip install tensorflow_addons"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "g12mQall-hes"
      },
      "outputs": [],
      "source": [
        "#@title ###**Load dependencies**.\n",
        "\n",
        "from google.colab import drive\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display\n",
        "import tensorflow as tf\n",
        "import numpy as np\n",
        "from sklearn.model_selection import train_test_split\n",
        "from tensorflow.keras.layers import LayerNormalization, MultiHeadAttention\n",
        "from tensorflow.keras.layers.experimental.preprocessing import TextVectorization\n",
        "import matplotlib.pyplot as plt\n",
        "import pandas as pd\n",
        "import csv\n",
        "import json\n",
        "import nltk\n",
        "from Bio import pairwise2\n",
        "from Bio.pairwise2 import format_alignment\n",
        "from Bio.Seq import Seq\n",
        "from Bio.Seq import CodonTable\n",
        "from tensorflow.keras.losses import Loss\n",
        "import tensorflow_addons as tfa\n",
        "import itertools as it\n",
        "import pandas as pd\n",
        "import dask.dataframe as dd\n",
        "from dask.diagnostics import ProgressBar"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "CwwkG_ly_Vpk"
      },
      "outputs": [],
      "source": [
        "#@title ###**Build tokenizer**.\n",
        "\n",
        "### FOR TOKENIZER, OPTIMALLY USE THE TEST DATASET SINCE IT IS SMALL\n",
        "# to test the trained model\n",
        "with open(\"/content/convergent_overlapping_gene_prediction/tokenizer_dataset/concat_pos_neg_114.json\") as concat:\n",
        "    concat_test_model = json.load(concat)\n",
        "\n",
        "with open(\"/content/convergent_overlapping_gene_prediction/tokenizer_dataset/overlap_114.json\") as overlap:\n",
        "    overlap_test_model = json.load(overlap)\n",
        "\n",
        "# Tokenization and text vectorization\n",
        "max_length = 103 #note, this should be the max len of the overlap, not input sequences\n",
        "vocab_size = 35\n",
        "\n",
        "\n",
        "# We must use the same tokenizer dataset as that used to train the original model\n",
        "def tokenize(sentences):\n",
        "    tokenizer = tf.keras.preprocessing.text.Tokenizer(\n",
        "        num_words=vocab_size, filters='')\n",
        "    tokenizer.fit_on_texts(sentences)\n",
        "    return tokenizer\n",
        "\n",
        "concat_tokenizer = tokenize(concat_test_model)\n",
        "overlap_tokenizer = tokenize(overlap_test_model)\n",
        "\n",
        "def vectorize(tokenizer, sentences):\n",
        "    seqs = tokenizer.texts_to_sequences(sentences)\n",
        "    return tf.keras.preprocessing.sequence.pad_sequences(seqs, maxlen=max_length, padding='post')\n",
        "\n",
        "concat_test_model_np = np.array(concat_test_model)\n",
        "overlap_test_model_np = np.array(overlap_test_model)\n",
        "\n",
        "concat_test_model_np_tokenizer = tokenize(concat_test_model_np)\n",
        "overlap_test_model_np_tokenizer = tokenize(overlap_test_model_np)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "wJtiidCd_BEb"
      },
      "outputs": [],
      "source": [
        "#@title ###**Load the overlap prediction model**.\n",
        "\n",
        "model_to_predict = tf.keras.models.load_model('/content/convergent_overlapping_gene_prediction/convergent_overlap_prediction_model')\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "Jw7N6M1h_BNE"
      },
      "outputs": [],
      "source": [
        "#@title ###**Prediction functions**.\n",
        "\n",
        "def predict_overlapping_sequence(trained_model, aa_sequences):\n",
        "\n",
        "    def translate_to_dna_with_all_options(aa_sequence: str) -> list:\n",
        "\n",
        "\n",
        "    # Codons and their frequencies for each amino acid based on the E. coli table\n",
        "        back_translation_code_with_all_options = {\n",
        "            'A': [('GCG', 0.27), ('GCT', 0.26), ('GCC', 0.26), ('GCA', 0.21)],\n",
        "            'C': [('TGC', 0.53), ('TGT', 0.47)],\n",
        "            'D': [('GAT', 0.63), ('GAC', 0.37)],\n",
        "            'E': [('GAA', 0.68), ('GAG', 0.32)],\n",
        "            'F': [('TTT', 0.58), ('TTC', 0.42)],\n",
        "            'G': [('GGC', 0.35), ('GGT', 0.32), ('GGG', 0.25), ('GGA', 0.08)],\n",
        "            'H': [('CAT', 0.56), ('CAC', 0.44)],\n",
        "            'I': [('ATT', 0.48), ('ATC', 0.39), ('ATA', 0.14)],\n",
        "            'K': [('AAA', 0.74), ('AAG', 0.26)],\n",
        "            'L': [('CTG', 0.43), ('CTT', 0.13), ('CTC', 0.13), ('TTA', 0.14), ('CTA', 0.07), ('TTG', 0.13)],\n",
        "            'M': [('ATG', 1.00)],\n",
        "            'N': [('AAC', 0.60), ('AAT', 0.40)],\n",
        "            'P': [('CCG', 0.52), ('CCA', 0.19), ('CCT', 0.16), ('CCC', 0.13)],\n",
        "            'Q': [('CAG', 0.66), ('CAA', 0.34)],\n",
        "            'R': [('CGT', 0.36), ('CGC', 0.36), ('CGG', 0.11), ('AGA', 0.08), ('AGG', 0.05), ('CGA', 0.04)],\n",
        "            'S': [('AGC', 0.24), ('TCC', 0.24), ('TCT', 0.17), ('TCG', 0.15), ('TCA', 0.14), ('AGT', 0.15)],\n",
        "            'T': [('ACC', 0.36), ('ACA', 0.28), ('ACG', 0.25), ('ACT', 0.11)],\n",
        "            'V': [('GTG', 0.46), ('GTT', 0.28), ('GTC', 0.15), ('GTA', 0.11)],\n",
        "            'W': [('TGG', 1.00)],\n",
        "            'Y': [('TAT', 0.59), ('TAC', 0.41)],\n",
        "            '*': [('TAA', 0.61), ('TGA', 0.30), ('TAG', 0.09)]\n",
        "        }\n",
        "\n",
        "        def choose_codon_based_on_frequency(codons):\n",
        "            \"\"\"\n",
        "            Randomly choose a codon based on the frequency distribution.\n",
        "\n",
        "            Args:\n",
        "            - codons (list): A list of tuples where each tuple contains a codon and its frequency\n",
        "\n",
        "            Returns:\n",
        "            - list: A list containing the chosen codon\n",
        "            \"\"\"\n",
        "            # Extract codon names and their frequencies\n",
        "            codon_names = [codon for codon, _ in codons]\n",
        "            codon_freqs = [freq for _, freq in codons]\n",
        "\n",
        "            # Normalize the frequencies to ensure they sum up to 1\n",
        "            total_frequency = sum(codon_freqs)\n",
        "            normalized_freqs = [freq/total_frequency for freq in codon_freqs]\n",
        "\n",
        "            # Randomly select a codon based on the frequency distribution\n",
        "            chosen_codon = np.random.choice(codon_names, p = normalized_freqs)\n",
        "\n",
        "            return [chosen_codon]\n",
        "\n",
        "        # For each amino acid in the sequence, choose a codon based on its frequency\n",
        "        list_of_list_of_codons = [choose_codon_based_on_frequency(back_translation_code_with_all_options[aa]) for aa in aa_sequence]\n",
        "\n",
        "        # Combine the codons to form the nucleotide sequence\n",
        "        list_of_combinations = [''.join(combination) for combination in it.product(*list_of_list_of_codons)]\n",
        "\n",
        "        return list_of_combinations\n",
        "\n",
        "    def translate(model, tokenizer, input_text):\n",
        "        input_vectorized = vectorize(tokenizer, [input_text])\n",
        "        prediction = model.predict(input_vectorized)[0]\n",
        "        predicted_indices = np.argmax(prediction, axis=-1)\n",
        "        #note that, we use the same tokenizer from the training data so that the output has the same tokens assigned\n",
        "        predicted_text = overlap_tokenizer.sequences_to_texts([predicted_indices])[\n",
        "            0].strip()\n",
        "        return predicted_text\n",
        "\n",
        "\n",
        "    def convert_chars_for_translated_overlap(string):\n",
        "        string = string.replace(\"no overlap\", \"#\", 1)  # Replace first occurrence of \"no overlap\" with \"#\"\n",
        "        string = string.replace(\"overlap\", \"#\", 1)  # Replace first occurrence of \"overlap\" with \"#\"\n",
        "        string = string.replace(\"no\", \"#\", 1)  # Replace first occurrence of \"no\" with \"#\"\n",
        "        if string == \"#\":\n",
        "            return string\n",
        "        new_string = \"\"\n",
        "        for char in string:\n",
        "            if char.lower() == \"b\":\n",
        "                new_string += \"A\"\n",
        "            elif char.lower() == \"j\":\n",
        "                new_string += \"G\"\n",
        "            elif char.lower() == \"o\":\n",
        "                new_string += \"T\"\n",
        "            elif char.lower() == \"u\":\n",
        "                new_string += \"C\"\n",
        "            else:\n",
        "                new_string += char\n",
        "        return new_string.replace(\" \", \"\")\n",
        "\n",
        "    # Running the above sequence through the model, using the tokenizer run on the data from the original model training process\n",
        "    translated_overlap = translate(trained_model, concat_test_model_np_tokenizer, aa_sequences)\n",
        "\n",
        "    overlap_output_from_model = convert_chars_for_translated_overlap(translated_overlap)\n",
        "\n",
        "    # Generate the rc to include in the next small df\n",
        "    reverse_complement_overlap = Seq(overlap_output_from_model)\n",
        "    reverse_complement_overlap = str(reverse_complement_overlap.reverse_complement())\n",
        "\n",
        "    # Create a DataFrame with two rows: original and reversed sequences\n",
        "    # Note, the rc sequence actually comes first here, followed by the model output overlap sequence\n",
        "    output_forward_reverse = pd.DataFrame({\"overlap_sequence\": [reverse_complement_overlap, overlap_output_from_model]})\n",
        "\n",
        "    # Splitting and collapsing the sequences\n",
        "    sequence_list = aa_sequences.split()\n",
        "    collapsed_sequence = ''.join(sequence_list)\n",
        "\n",
        "    # Splitting the sequence at the asterisk and keeping the asterisk\n",
        "    split_sequences_with_asterisk = [seq + '*' for seq in collapsed_sequence.split('*') if seq]\n",
        "\n",
        "    # Creating a dataframe from the sequences\n",
        "    df_sequences = pd.DataFrame(split_sequences_with_asterisk, columns=[\"amino_acid_sequence\"])\n",
        "\n",
        "    # Adding the back translated dna sequences to the data frame.\n",
        "    df_sequences['nt_sequence'] = df_sequences['amino_acid_sequence'].apply(lambda aa_seq: translate_to_dna_with_all_options(aa_seq)[0])\n",
        "\n",
        "    df_sequences[\"overlap_seq\"] = output_forward_reverse\n",
        "\n",
        "    #here, we are attaching the overlap sequence to the known coding sequence, generated above. It is\n",
        "    #added at the location based on the length of the overlap_seq\n",
        "    def modify_sequence(df):\n",
        "        # Check if 'overlap_seq' column exists in the DataFrame\n",
        "        if 'overlap_seq' not in df.columns:\n",
        "            raise ValueError(\"DataFrame must contain 'overlap_seq' column.\")\n",
        "\n",
        "        # Calculate the length of the overlap sequence\n",
        "        df['overlap_length'] = df['overlap_seq'].apply(len)\n",
        "\n",
        "        # Modify the sequences\n",
        "        df['modified_sequence'] = df.apply(lambda row: row['nt_sequence'][:-row['overlap_length']] + row['overlap_seq'], axis=1)\n",
        "\n",
        "        return df\n",
        "\n",
        "    modify_sequence(df_sequences)\n",
        "\n",
        "    # Custom function to translate using Seq\n",
        "    def translate_with_seq(sequence):\n",
        "        coding_dna = Seq(sequence)\n",
        "        return str(coding_dna.translate())\n",
        "\n",
        "    # Apply the custom function to the 'nt_sequence' column\n",
        "    df_sequences['translated_sequence'] = df_sequences['modified_sequence'].apply(translate_with_seq)\n",
        "\n",
        "    #print(compare_aa)\n",
        "    return(df_sequences)\n",
        "\n",
        "\n",
        "def find_matching_sequence(model_to_predict, formatted_aa_seq, max_attempts, _counter=[0]):\n",
        "    _counter[0] += 1\n",
        "    print(f\"Parsing row {_counter[0]}...\")\n",
        "\n",
        "    attempts = 0\n",
        "    match_found = False\n",
        "    matching_dataframe = None\n",
        "\n",
        "    while attempts < max_attempts:\n",
        "        attempts += 1\n",
        "\n",
        "        try:\n",
        "            result_df = predict_overlapping_sequence(model_to_predict, formatted_aa_seq)\n",
        "        except CodonTable.TranslationError as e:\n",
        "            print(f\"Error: {e}\")\n",
        "             # Print the \"overlap_length\" at the end of each iteration\n",
        "            print(f\"Attempt {attempts}... Overlap Length: {matching_dataframe['overlap_length'].iloc[0] if matching_dataframe is not None else 'N/A'}\")\n",
        "            continue\n",
        "\n",
        "        # Check if the \"nt_sequence\" and \"translated_sequence\" columns match in both rows\n",
        "        if (result_df['amino_acid_sequence'].iloc[0] == result_df['translated_sequence'].iloc[0] and\n",
        "            result_df['amino_acid_sequence'].iloc[1] == result_df['translated_sequence'].iloc[1]):\n",
        "            match_found = True\n",
        "            matching_dataframe = result_df\n",
        "            break\n",
        "\n",
        "    if match_found:\n",
        "         # Print the \"overlap_length\" at the end of each iteration\n",
        "        print(f\"Attempt {attempts}... Overlap Length: {matching_dataframe['overlap_length'].iloc[0] if matching_dataframe is not None else 'N/A'}\")\n",
        "        return matching_dataframe\n",
        "    else:\n",
        "        print(\"There is no predicted significant overlap\")\n",
        "        return None\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "qE3IHKc7q_qi"
      },
      "outputs": [],
      "source": [
        "#@title ###**Process the input amino acid sequences**. Note, both sequences must be at least 33 amino acids long\n",
        "\n",
        "sequence_1 = \"MLALAPVIAAVGEGRESMGGWKKIYRAALNGELEPMPGPDLPPPPPWSTVIHELWDASRTLPWTHLRGPLAPAMLRDHHDQALAVASAQPPG\" #@param {type:\"string\"}\n",
        "sequence_2 = \"MIIWFVAPGTRVLLLPVVALRRGILQKAPMERAWWTATLVLVAMHATDLPLFDSRLNILGWTLLAGLAAFNQEVGQMPQPRPDRDGPAASPEPKAP\" #@param {type:\"string\"}\n",
        "\n",
        "def process_sequences(aa_seq_1, aa_seq_2):\n",
        "\n",
        "    if len(aa_seq_1) < 33:\n",
        "        return \"Error: sequence_1 is not long enough to process. The minimum length is 33 amino acids.\"\n",
        "    if len(aa_seq_2) < 33:\n",
        "        return \"Error: sequence_2 is not long enough to process. The minimum length is 33 amino acids.\"\n",
        "\n",
        "    # Keep only the final 34 amino acids of each sequence\n",
        "    aa_seq_1_trimmed = aa_seq_1[-34:]\n",
        "    aa_seq_2_trimmed = aa_seq_2[-34:]\n",
        "\n",
        "    # Replace the first amino acid with an 'M'\n",
        "    aa_seq_1_processed = 'M' + aa_seq_1_trimmed[1:]\n",
        "    aa_seq_2_processed = 'M' + aa_seq_2_trimmed[1:]\n",
        "\n",
        "    # Add an asterisk after each sequence\n",
        "    aa_seq_1_processed += '*'\n",
        "    aa_seq_2_processed += '*'\n",
        "\n",
        "    # Concatenate the two sequences\n",
        "    concatenated_seq = aa_seq_1_processed + aa_seq_2_processed\n",
        "\n",
        "    # Add a space between every character\n",
        "    final_seq = ' '.join(concatenated_seq)\n",
        "\n",
        "    # Return the processed sequence\n",
        "    return final_seq\n",
        "\n",
        "concatenated_sequence = process_sequences(sequence_1, sequence_2)\n",
        "\n",
        "# Check if the result is an error message or the processed sequence, and print accordingly\n",
        "if \"Error\" in concatenated_sequence:\n",
        "    print(concatenated_sequence)\n",
        "else:\n",
        "    print(\"The input amino acid sequences have been processed and concatenated: \")\n",
        "    print(concatenated_sequence)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "UwDlAZaYB3MU"
      },
      "outputs": [],
      "source": [
        "#@title ###**Predict a convergent overlapping sequence**.\n",
        "\n",
        "formatted_aa_seq = concatenated_sequence\n",
        "\n",
        "inference_rounds = \"10\" # @param [\"10\", \"100\", \"200\", \"300\", \"500\", \"1000\"] {type:\"string\"}\n",
        "\n",
        "inference_rounds = int(inference_rounds)\n",
        "\n",
        "testing_df = find_matching_sequence(model_to_predict, formatted_aa_seq, inference_rounds)\n",
        "\n",
        "testing_df"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
